### PR TITLE
[meshcop] implement commissioning UDP proxy

### PR DESCRIPTION
**Pull request by bukepo**
This PR implements the Mesh Commissioning UDP proxy.

### CODE REVIEW

**Comment by turon** [comment]
🥇 Yay!  I like this feature.  Is there any new cli to test or drive the feature with?

File: src/core/meshcop/border_agent.cpp [hunk 1]
```diff
@@ -363,22 +370,49 @@ bool BorderAgent::HandleProxyMessage(const Message &aMessage, const Ip6::Message
...
+        UdpEncapsulationTlv tlv;
+        uint16_t            udpLength = aMessage.GetLength() - aMessage.GetOffset();
+
+        tlv.Init();
+        tlv.SetSourcePort(aMessageInfo.GetPeerPort());
+        tlv.SetDestinationPort(aMessageInfo.GetSockPort());
+        tlv.SetUdpLength(udpLength);
+        SuccessOrExit(error = message->SetLength(tlv.GetLength()));
+        SuccessOrExit(error = message->Append(&tlv, sizeof(tlv)));
+        VerifyOrExit(aMessage.CopyTo(sizeof(tlv), 0, udpLength, *message) == udpLength, error = OT_ERROR_NO_BUFS);
```
Comment (gjc13) [inline src/core/meshcop/border_agent.cpp]: aMessage.CopyTo(aMessage.GetOffset(), sizeof(tlv), udpLength, *message)

**Comment by bukepo** [comment]
@turon we are going to use this feature in border router soon.

File: src/core/meshcop/border_agent.cpp (hunk 1 above)
Comment (bukepo) [reply src/core/meshcop/border_agent.cpp]: Thanks.

File: src/core/meshcop/border_agent.cpp [hunk 2]
```diff
@@ -252,11 +312,110 @@ BorderAgent::BorderAgent(Instance &aInstance)
...
+
+    {
+        UdpEncapsulationTlv tlv;
+        uint16_t            udpLength = aMessage.GetLength() - aMessage.GetOffset();
+
+        tlv.Init();
+        tlv.SetSourcePort(aMessageInfo.GetPeerPort());
+        tlv.SetDestinationPort(aMessageInfo.GetSockPort());
+        tlv.SetUdpLength(udpLength);
+        SuccessOrExit(error = message->SetLength(tlv.GetLength()));
```
Comment (gjc13) [inline src/core/meshcop/border_agent.cpp]: message->SetLength(sizeof(tlv) + udpLength);

**Comment by gjc13** [comment]
Works in border router application.
LGTM
:+1:

File: src/core/meshcop/border_agent.cpp [hunk 3]
```diff
@@ -166,30 +192,53 @@ void BorderAgent::HandleCoapResponse(void *               aContext,
...
+    SuccessOrExit(error = borderAgent.ForwardToCommissioner(header, *message));
 
 exit:
     netif.GetInstance().GetHeap().Free(&forwardContext);
+
+    if (error != OT_ERROR_NONE)
+    {
+        otLogWarnMeshCoP(GetInstance(), "Commissioner request[%hu] failed: %s", forwardContext.GetMessageId(),
+                         otThreadErrorToString(aResult));
+        forwardContext.ToHeader(header, CoapCodeFromError(error));
```
Comment (jwhui) [inline src/core/meshcop/border_agent.cpp]: Make sure to free `forwardContext` after this usage?

File: src/core/meshcop/meshcop_tlvs.hpp [hunk 4]
```diff
@@ -178,6 +181,41 @@ class Tlv : public ot::Tlv
...
+
+    /**
+     * This method returns a pointer to the next TLV.
+     *
+     * @returns A pointer to the next TLV.
+     *
+     */
+    Tlv *GetNext(void) { return static_cast<Tlv *>(ot::ExtendedTlv::GetNext()); }
+
+    const Tlv *GetNext(void) const { return static_cast<const Tlv *>(ot::Tlv::GetNext()); }
```
Comment (jwhui) [inline src/core/meshcop/meshcop_tlvs.hpp]: Are these `GetNext()` methods used anywhere?

File: src/core/meshcop/meshcop_tlvs.hpp [hunk 5]
```diff
@@ -2175,6 +2213,136 @@ class VendorStackVersionTlv : public Tlv
...
+ */
+OT_TOOL_PACKED_BEGIN
+class IPv6AddressTlv : public Tlv
+{
+public:
+    /**
+     * Default constructor.
+     *
+     */
+    IPv6AddressTlv(void) {}
```
Comment (jwhui) [inline src/core/meshcop/meshcop_tlvs.hpp]: Is this user-specified constructor necessary?

File: src/core/meshcop/meshcop_tlvs.hpp [hunk 6]
```diff
@@ -2175,6 +2213,136 @@ class VendorStackVersionTlv : public Tlv
...
+ */
+OT_TOOL_PACKED_BEGIN
+class UdpEncapsulationTlv : public ExtendedTlv
+{
+public:
+    /**
+     * Default constructor.
+     *
+     */
+    UdpEncapsulationTlv(void) {}
```
Comment (jwhui) [inline src/core/meshcop/meshcop_tlvs.hpp]: Is this user-specified constructor necessary?

File: src/core/thread/mle.hpp [hunk 7]
```diff
@@ -950,7 +966,7 @@ class Mle : public InstanceLocator
     /**
      * This method returns the Service Aloc corresponding to a Service ID.
      *
-     * @param[in]  aAloc16  The Servicer ID value.
+     * @param[in]  aServiceId  The Servicer ID value.
```
Comment (jwhui) [inline src/core/thread/mle.hpp]: typo: `Servicer` -> `Service`

File: include/openthread/udp.h [hunk 8]
```diff
@@ -53,6 +53,26 @@ extern "C" {
  *
  */
 
+/**
+ * This callback allows OpenThread to provide specific handler for certain UDP message.
```
Comment (jwhui) [inline include/openthread/udp.h]: nit: "to provide specific handlers for certain UDP messages"

File: src/core/meshcop/border_agent.cpp [hunk 9]
```diff
@@ -61,21 +61,46 @@ class ForwardContext
      *
      * @param[in]   aBorderAgent    A reference to the border agent.
      * @param[in]   aHeader         A reference to the request header.
-     * @param[in]   aSeparate       Whether this request should be responded separately.
+     * @param[in]   aPetition       Whether this request is leader petition.
```
Comment (jwhui) [inline src/core/meshcop/border_agent.cpp]: nit: Both LEAD_PET and COMM_PET messages exist.  Should we simplify to just say "petition"?
e.g. "is leader petition" -> "is a petition"

File: src/core/meshcop/border_agent.cpp [hunk 10]
```diff
@@ -252,11 +312,116 @@ BorderAgent::BorderAgent(Instance &aInstance)
...
+        otLogWarnMeshCoP(GetInstance(), "Failed to send proxy stream: %s", otThreadErrorToString(error));
+
+        if (message != NULL)
+        {
+            message->Free();
+        }
+    }
+}
+
+bool BorderAgent::HandleProxyMessage(const Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
```
Comment (jwhui) [inline src/core/meshcop/border_agent.cpp]: nit: rename to `HandleProxyReceive()`?

File: src/core/meshcop/border_agent.cpp [hunk 11]
```diff
@@ -252,11 +312,116 @@ BorderAgent::BorderAgent(Instance &aInstance)
...
+    }
+
+    SuccessOrExit(error = netif.GetCoapSecure().SendMessage(*message, netif.GetCoapSecure().GetPeerMessageInfo()));
+
+    otLogInfoMeshCoP(GetInstance(), "Sent to commissioner on %s", OT_URI_PATH_PROXY_RX);
+
+exit:
+    if (message != NULL && error != OT_ERROR_NONE)
+    {
+        message->Free();
```
Comment (jwhui) [inline src/core/meshcop/border_agent.cpp]: Add log on error?

File: src/core/meshcop/border_agent.cpp [hunk 12]
```diff
@@ -252,11 +312,116 @@ BorderAgent::BorderAgent(Instance &aInstance)
...
+
+        tlv.Init();
+        tlv.SetSourcePort(aMessageInfo.GetPeerPort());
+        tlv.SetDestinationPort(aMessageInfo.GetSockPort());
+        tlv.SetUdpLength(udpLength);
+        SuccessOrExit(error = message->Append(&tlv, sizeof(tlv)));
+
+        offset = message->GetLength();
+        SuccessOrExit(error = message->SetLength(offset + udpLength));
+        VerifyOrExit(aMessage.CopyTo(aMessage.GetOffset(), offset, udpLength, *message) == udpLength,
```
Comment (jwhui) [inline src/core/meshcop/border_agent.cpp]: assert here instead?  `SetLength()` should allocate the appropriate message buffer.

File: src/core/thread/mle.hpp [hunk 13]
```diff
@@ -855,7 +856,22 @@ class Mle : public InstanceLocator
      * @retval OT_ERROR_DETACHED  The Thread interface is not currently attached to a Thread Partition.
      *
      */
-    otError GetLeaderAloc(Ip6::Address &aAddress) const;
+    otError GetLeaderAloc(Ip6::Address &aAddress) const { return GetAlocAddress(aAddress, kAloc16Leader); }
+
+    /**
+     * This method retrieves the Commissioner's ALOC.
```
Comment (jwhui) [inline src/core/thread/mle.hpp]: nit: "retrieves" -> "computes"

**Pull request merged**